feat(oxfmt): Support `.oxfmtrc.json(c)` config file

leaysgur opened the pull request:
Fixes #13611

- Support config file with JSON(C) format
- Automatically find the nearest one from `cwd`
  - Do not merge, just use the first one
  - Search only `.oxfmtrc.json`(not `.jsonc`)
- Use `-c` or `--config` to specify path

Example `.oxfmtrc.json`:

```jsonc
{
  "indentStyle": "space",
  "indentWidth": 2,
  "lineWidth": 80,
  "lineEnding": "lf",
  "quoteStyle": "single",
  "jsxQuoteStyle": "double",
  "quoteProperties": "as-needed",
  "trailingCommas": "all",
  "semicolons": "always",
  "arrowParentheses": "always",
  "bracketSpacing": true,
  "bracketSameLine": false,
  "attributePosition": "auto",
  "expand": "auto",
  "experimentalOperatorPosition": "before",
  "experimentalSortImports": {
    "partitionByNewline": false,
    "partitionByComment": false,
    "sortSideEffects": false,
    "order": "asc",
    "ignoreCase": true
  }
}
```

We may provide tools like `oxfmt-migrate` to convert `.prettierrc`.

leaysgur commented:
* **#14398** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/14398?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/14398?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=14398).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Sysix reviewed (approved):
looks good for the server 👍 thank you

camc314 commented on apps/oxfmt/src/format.rs:
```diff
@@ -156,6 +175,50 @@ impl FormatRunner {
     }
 }
 
+const DEFAULT_CONFIG_NAMES: &[&str] = &[".oxfmtrc.json", ".oxfmtrc.jsonc"];
```
@Sysix should we stick with just one config name? `.oxfmtrc.jsonc` ?

leaysgur replied:
Thank you both. I'll align it (only search `.json`) for now.

Dunqing reviewed (changes requested):
Looks good! There is a tiny thing that can be simplified a little.

leaysgur committed:
feat(oxfmt): Support `.oxfmtrc.json(c)` config file (#14398)

Fixes #13611

- Support config file with JSON(C) format
- Automatically find the nearest one from `cwd`
  - Do not merge, just use the first one
  - Search only `.oxfmtrc.json`(not `.jsonc`)
- Use `-c` or `--config` to specify path

Example `.oxfmtrc.json`:

```jsonc
{
  "indentStyle": "space",
  "indentWidth": 2,
  "lineWidth": 80,
  "lineEnding": "lf",
  "quoteStyle": "single",
  "jsxQuoteStyle": "double",
  "quoteProperties": "as-needed",
  "trailingCommas": "all",
  "semicolons": "always",
  "arrowParentheses": "always",
  "bracketSpacing": true,
  "bracketSameLine": false,
  "attributePosition": "auto",
  "expand": "auto",
  "experimentalOperatorPosition": "before",
  "experimentalSortImports": {
    "partitionByNewline": false,
    "partitionByComment": false,
    "sortSideEffects": false,
    "order": "asc",
    "ignoreCase": true
  }
}
```

We may provide tools like `oxfmt-migrate` to convert `.prettierrc`.
